onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc

```diff
@@ -0,0 +1,356 @@
+  // this could be replaced with ANEURALNETWORKS_BATCH_MATMUL when that is more widely supported
+
+  // assuming A and B have at least three dimensions and the same leading dimensions other than the last two
+
+  const auto& inputs = node_unit.Inputs();
+
+  Shape a_shape;
+  ORT_RETURN_IF_NOT(GetShape(inputs[0].node_arg, a_shape), "Failed to get A's shape.");
+
+  Shape b_shape;
+  ORT_RETURN_IF_NOT(GetShape(inputs[1].node_arg, b_shape), "Failed to get B's shape.");
+
+  const std::string& a = inputs[0].node_arg.Name();
+  const std::string& b = inputs[1].node_arg.Name();
+
+  const std::string& output = node_unit.Outputs()[0].node_arg.Name();
+
+  std::vector<std::string> gemm_a_inputs{a};
+  std::vector<std::string> gemm_b_inputs{b};
+
+  const auto m = a_shape[a_shape.size() - 2],
+             k = a_shape[a_shape.size() - 1],
+             n = b_shape[b_shape.size() - 1];
+
+  const bool reshape_leading_dimensions = a_shape.size() > 3;
+  const auto batch_size = ShapeSize(a_shape, 0, a_shape.size() - 2);
+
+  auto add_reshape = [&model_builder](const std::string& input, const Shape& new_shape,
+                                      const std::string& output) -> Status {
+    const std::string new_shape_name = model_builder.GetUniqueName(input + "/new_shape");
+    std::vector<int32_t> new_shape_i32{};
+    new_shape_i32.reserve(new_shape.size());
+    std::transform(new_shape.begin(), new_shape.end(), std::back_inserter(new_shape_i32),
+                   [](uint32_t d) { return gsl::narrow<int32_t>(d); });
+    ORT_RETURN_IF_ERROR(AddNnapiReshape(model_builder, input, new_shape_name, new_shape_i32, output, nullptr));
+    return Status::OK();
+  };
+
+  auto add_reshape_generate_output = [&model_builder, &add_reshape](const std::string& input, const Shape& new_shape,
+                                                                    std::string& output) -> Status {
+    std::string reshaped = model_builder.GetUniqueName(input + "/reshaped");
+    ORT_RETURN_IF_ERROR(add_reshape(input, new_shape, reshaped));
+    output = std::move(reshaped);
+    return Status::OK();
+  };
+
+  // collapse leading dimensions to a single one
+  if (reshape_leading_dimensions) {
+    const Shape a_new_shape_value = {batch_size, m, k},
+                b_new_shape_value = {batch_size, k, n};
+    std::string a_reshaped, b_reshaped;
+
+    ORT_RETURN_IF_ERROR(add_reshape_generate_output(gemm_a_inputs.front(), a_new_shape_value, a_reshaped));
+    gemm_a_inputs.front() = a_reshaped;
+
+    ORT_RETURN_IF_ERROR(add_reshape_generate_output(gemm_b_inputs.front(), b_new_shape_value, b_reshaped));
+    gemm_b_inputs.front() = b_reshaped;
+  }
+
+  // transpose B
+  {
+    const std::string b_new_perm = model_builder.GetUniqueName(b + "/new_perm"),
+                      b_transposed = model_builder.GetUniqueName(b + "/transposed");
+    ORT_RETURN_IF_ERROR(AddNnapiTranspose(model_builder, gemm_b_inputs.front(), b_new_perm, {0, 2, 1}, b_transposed));
```

Comment:
> {0, 2, 1}

nit question:  just trying to understand the implementation here, what is this permutation for transpose based on?